From 45e7665568e4d94d96e1c7269ed0782e82f390c1 Mon Sep 17 00:00:00 2001
From: kk <kaiwalya.khasnis57@gmail.com>
Date: Wed, 14 Oct 2026 11:33:31 +0900
Subject: [PATCH 2/2] bullhead: sepolicy for RGB LED HAL and framework service

- hal_rgbled attribute triple, declared by hand (8.1 has no hal_attribute())
- hal_rgbled_default domain for the service, with access to sysfs_leds
- label the qpnp LED class devices as sysfs_leds
- rgbled binder service for system_server, reachable from apps
---
 sepolicy/file_contexts         |  3 +++
 sepolicy/genfs_contexts        |  3 +++
 sepolicy/hal_rgbled.te         | 15 +++++++++++++++
 sepolicy/hal_rgbled_default.te |  5 +++++
 sepolicy/hwservice.te          |  1 +
 sepolicy/hwservice_contexts    |  1 +
 sepolicy/service.te            |  3 +++
 sepolicy/service_contexts      |  1 +
 sepolicy/system_server.te      |  3 +++
 9 files changed, 35 insertions(+)
 create mode 100644 sepolicy/hal_rgbled.te
 create mode 100644 sepolicy/hal_rgbled_default.te
 create mode 100644 sepolicy/hwservice.te
 create mode 100644 sepolicy/hwservice_contexts

diff --git a/sepolicy/file_contexts b/sepolicy/file_contexts
index 314c7cd..0e7eb0a 100644
--- a/sepolicy/file_contexts
+++ b/sepolicy/file_contexts
@@ -185,3 +185,6 @@
 /persist/wlan_mac\.bin             u:object_r:persist_wifi_file:s0
 /persist/sensorcal\.json           u:object_r:persist_sensortool_file:s0
 /persist/calibration(/.*)?         u:object_r:persist_camera_file:s0
+
+# RGB LED HAL
+/system/bin/hw/vendor\.lineage\.rgbled@1\.0-service\.bullhead  u:object_r:hal_rgbled_default_exec:s0
diff --git a/sepolicy/genfs_contexts b/sepolicy/genfs_contexts
index cfbde95..e46362a 100644
--- a/sepolicy/genfs_contexts
+++ b/sepolicy/genfs_contexts
@@ -9,3 +9,6 @@ genfscon proc /sys/kernel/sched_small_task       u:object_r:proc_kernel_sched:s0
 genfscon proc /sys/kernel/sched_boost            u:object_r:proc_kernel_sched:s0
 genfscon proc /debug/fwdump                      u:object_r:proc_wifi_dbg:s0
 genfscon proc /debugdriver/driverdump            u:object_r:proc_wifi_dbg:s0
+
+# Notification RGB LED (red/green/blue class devices)
+genfscon sysfs /devices/soc.0/leds-qpnp-25/leds u:object_r:sysfs_leds:s0
diff --git a/sepolicy/hal_rgbled.te b/sepolicy/hal_rgbled.te
new file mode 100644
index 0000000..b576924
--- /dev/null
+++ b/sepolicy/hal_rgbled.te
@@ -0,0 +1,15 @@
+# RGB LED HAL (vendor.lineage.rgbled@1.0). 8.1 has no hal_attribute() macro,
+# so the attribute triple is declared by hand.
+attribute hal_rgbled;
+attribute hal_rgbled_client;
+attribute hal_rgbled_server;
+
+binder_call(hal_rgbled_client, hal_rgbled_server)
+binder_call(hal_rgbled_server, hal_rgbled_client)
+
+add_hwservice(hal_rgbled_server, hal_rgbled_hwservice)
+allow hal_rgbled_client hal_rgbled_hwservice:hwservice_manager find;
+
+allow hal_rgbled sysfs_leds:lnk_file read;
+allow hal_rgbled sysfs_leds:file rw_file_perms;
+allow hal_rgbled sysfs_leds:dir r_dir_perms;
diff --git a/sepolicy/hal_rgbled_default.te b/sepolicy/hal_rgbled_default.te
new file mode 100644
index 0000000..30e6ccf
--- /dev/null
+++ b/sepolicy/hal_rgbled_default.te
@@ -0,0 +1,5 @@
+type hal_rgbled_default, domain;
+hal_server_domain(hal_rgbled_default, hal_rgbled)
+
+type hal_rgbled_default_exec, exec_type, file_type;
+init_daemon_domain(hal_rgbled_default)
diff --git a/sepolicy/hwservice.te b/sepolicy/hwservice.te
new file mode 100644
index 0000000..2ac49b7
--- /dev/null
+++ b/sepolicy/hwservice.te
@@ -0,0 +1 @@
+type hal_rgbled_hwservice, hwservice_manager_type;
diff --git a/sepolicy/hwservice_contexts b/sepolicy/hwservice_contexts
new file mode 100644
index 0000000..ece9652
--- /dev/null
+++ b/sepolicy/hwservice_contexts
@@ -0,0 +1 @@
+vendor.lineage.rgbled::IRgbLed                    u:object_r:hal_rgbled_hwservice:s0
diff --git a/sepolicy/service.te b/sepolicy/service.te
index 0e757fa..62b3d76 100644
--- a/sepolicy/service.te
+++ b/sepolicy/service.te
@@ -1,3 +1,6 @@
 type per_mgr_service,           service_manager_type;
 type atfwd_service,             service_manager_type;
 type cne_service,               service_manager_type;
+
+# RGB LED framework service
+type rgbled_service, app_api_service, system_server_service, service_manager_type;
diff --git a/sepolicy/service_contexts b/sepolicy/service_contexts
index 13c8608..9b452bd 100644
--- a/sepolicy/service_contexts
+++ b/sepolicy/service_contexts
@@ -1,3 +1,4 @@
 AtCmdFwd                                  u:object_r:atfwd_service:s0
 vendor.qcom.PeripheralManager             u:object_r:per_mgr_service:s0
 cneservice                                u:object_r:cne_service:s0
+rgbled                                    u:object_r:rgbled_service:s0
diff --git a/sepolicy/system_server.te b/sepolicy/system_server.te
index 8b0f1bf..63d7bf6 100644
--- a/sepolicy/system_server.te
+++ b/sepolicy/system_server.te
@@ -40,3 +40,6 @@ allowxperm system_server self:socket ioctl msm_sock_ipc_ioctls;
 
 # system_server calls the power HAL, which reads debugfs_rpm
 allow system_server debugfs_rpm:file { read open getattr };
+
+# RGB LED HAL
+hal_client_domain(system_server, hal_rgbled)
